### DISCUSSION TITLE
[FEAT] Frogs

### DISCUSSION THREAD

**Discussion by tiffanydys** [sunflower-land/sunflower-land#1217] [Ideas]
## Art
![banner](https://user-images.githubusercontent.com/18549210/178298257-5814a849-105c-41b4-ae43-d82a9a3657ef.png)
courtesy of @paluras and @VadimasVP

### Preview of the minted frog: (to be revealed later after all frogs mint out)
![frog_random](https://user-images.githubusercontent.com/18549210/178312365-18936cc4-33dd-4d5f-89fe-a12c3080df90.gif)

### Sample frog: (After frogs mint out)
![3](https://user-images.githubusercontent.com/18549210/178312828-dacaecda-7f8d-40d1-984c-61e62a0e1532.png)

### Generation of Frogs
For this, I've decided to use [HashLips' Art Engine](https://github.com/HashLips/hashlips_art_engine). 
What it does is combining each layer of your art into thousands of uniquely generated artworks.
It also generates the metadata for each frog, so you don't have to.

### Rarities
HashLips' Art Engine also supports rarities. You will only need to attach a ``#`` followed by the rarity number on the png image name like so: ``Wizard#100.png``. The **bigger** the number, the **more often** it will occur on each generated frog.

**Base**
<img width="305" alt="Screen Shot 2022-07-12 at 12 20 04 AM" src="https://user-images.githubusercontent.com/18549210/178311707-78ea9f08-d41d-46a2-83b0-8274c93f71e1.png">

**Hat**
<img width="308" alt="Screen Shot 2022-07-12 at 12 19 52 AM" src="https://user-images.githubusercontent.com/18549210/178311741-966231a6-9255-45b0-9363-7809c1e9d140.png">

**Float**
<img width="299" alt="Screen Shot 2022-07-12 at 1 18 44 AM" src="https://user-images.githubusercontent.com/18549210/178321292-d1b89a3a-07ca-4ae9-8fdc-54e80a1c0f7d.png">

### Sample of Generated Frogs Using HashLips' Art Engine:
![image](https://user-images.githubusercontent.com/18549210/178314696-066b98d6-bdd7-4c59-94ab-85b40eeb8789.png)

### Sample Contract Collection Deployed on Testnet(Dummy Contract - Unfinished and NOT official)
<img width="359" alt="Screen Shot 2022-07-11 at 11 15 41 PM" src="https://user-images.githubusercontent.com/18549210/178314404-5848b6bc-9d78-42c9-b6a4-f4fe2a0d31ff.png">

### Front-End

- The frog will be available to mint using the `Merchant` found on Goblin Land. 
- PR is ready and will be submitted by me after we've planned it out in this discussion.
- 1 Farm will be showing only 1 Frog (even if you have many)

![image](https://user-images.githubusercontent.com/18549210/178315207-d7ee7aa4-9cdd-48b6-9bdb-101dba1ea8c8.png)

### Back-End API
- Mint would probably cost around 50-100 SFL. (Up to the SFL Team to choose)
- Limit 1 mint per farm with a cool down of 7 days.
- Supply would be around 300-500 (it depends what the team wants)

### Smart Contract

I will be submitting a PR for this as well, just need to finalize some information with the team.
The contract will include the following:

- Collection: Sunflower Land Frogs
- contract SunflowerLandFrogs is GameOwner, ERC721Enumerable

functions:
- setBaseUri (will most likely be https://sunflower-land.com/play/nfts/frogs)
- gameMint
- gameBurn
- gameTransfer

### SFL Integration
What I'm thinking is that each unique frog will follow the ERC721 format when minted, and will be included in SunflowerLandInventory. After detecting you have one it will display on your farm. (Only one frog on each farm - could be more what do you think?) Also with Land Expansion incoming, I'm curious to know how the other NFTs will be displayed. 

If you got any questions please feel free to comment them down and we'll discuss 😄
Thank you!
 
Made by: @DonatasDee @VadimasVP @paluras @tiffanydys

**Comment by adamhannigan** [comment]
Awesome work @tiffanydys and all of the artists ❤️ 

One technical challenge is the burning of SFL from a farm. This becomes difficult since players can be in the middle of playing the game, crafting, buying, selling etc. It essentially becomes a juggle of ensuring that the on chain SFL is aligned with the off chain SFL - one wrong move opens the door for vulnerabilities such as crafting a frog with SFL and then quickly spending that SFL in a different in game tab. 

The way we currently solve this is by integrating session logic throughout our smart contracts and our API architecture. For community features this won't scale very well as we need to build custom logic and maintain it for each feature.

Where I see community features thriving is in cases where the developers/artists have control of the architecture, smart contracts and direction which means less blockers and more room to innovate. 


Instead of waiting for the team to build out the session logic, I would suggest that the SFL required to mint a frog actually comes from the player's personal wallet. In this scenario the frog feature would not need to hit our game's API, just the smart contracts you deploy. This is great for decentralisation and decoupling features that could bloat our architecture.

This could function as:

1. Approval for your smart contract to transfer SFL
2. Mint frog (which burns SFL) and send to farm ID


Fundamentally this would evolve into you and your team owning this product. You could set Open Sea royalties if desired on the smart contracts you deploy. Hopefully this is a benefit and helps incentives the builders to keep maintaining and improving the feature ❤️

**Reply by AL-Simmons92** [reply]
Very Cool!!! LFG

**Reply by eodestroyer** [reply]
Love this!! Would be nice if they were like chickens and spawned tad poles when fed.  Then there is a chance to spawn a mutant frog that may benefit the pond ecosystem.

**Reply by AL-Simmons92** [reply]
Make it a 1/1000 drop  like mutant chicken but with a 10,000 limit. To increase rarity of mutants.

**Reply by eodestroyer** [reply]
A limit would be nice.

**Reply by alphacruze** [reply]
> A limit would be nice.

You want to introduce a production based resource that has a supply of 300? that's ridiculous, then it should not be an exclusive like T3's with super low supply, it should be obtainable by every player, not just a select few who happen to be lucky enough to mint it, instead it should then be like how Chickens can be owned by every player, however sticking with it's limit of 1 per farm, much like the idea of Slimes, where there's 1 per farm, however unlike slimes that are land based, frogs will be Water based, with the increased supply to cost per frog should then also be decreased to something like 10 - 25 per frog. 

WIth rarity having no real value except appearance or perhaps like slimes, the ability to feed the frogs, higher rarity frogs can be fed more thus they produce more in the same amount of time, naturally in a sense upkeep for maximum returns is much much higher.

An Example, a Common frog can only be fed 10 of resource X, while A legendary can be fed 100 of Resource X.
Both will produce X in let's say 24 hours. However The legendary will produce 10x what the common produces as it can be fed more.

Now this does create a lot of extra problems in the form of a new resource and all that, but that's only if decide to go based on a production type Animal, if it's just cosmetics, well having a unlimited supply costing 10 SFL compared to a limited supply of 300 costing 100 SFL is a lot better

As the current model can only burn 300 x 100 = 30 000 SFL, which is insignificant considering there's around 9 Mil SFL that is unburnt(which will continue indefinitely, after all, you receive more for selling your crops than you pay for them... so i mean pretty obvious right?, well whats not obvious is that we likely to always see just around slightly more than half of the SFL supply being burnt)

Now back to my point if the supply is unlimited, we will see a lot more people getting frogs so the potential burn is ALOT higher than just 30k... example 10k players get frogs at 10 SFL, we see 100k SFL burnt... if that goes up to 50k, well 500k burnt

At that point it just comes down to how well Frogs are received.

**Comment by celpower** [comment]
I love the idea of the frogs and how they can be different and you have no idea which one you will get, with that said, they need to have utility. I hope SFL doesn't become an expensive sticker book/decoration NFTs.

**Reply by Yougottabekiddingme123** [reply]
Agreed. All frogs should have some sort of utility.

**Comment by Yougottabekiddingme123** [comment]
Utility should be given to all frogs. Utility can scale from 1% to 5%

**Comment by Theadra** [comment]
I love the frogs <3 

Sharing an idea; Each frog can represent an item or process, example, a fast food clerk themed frog gives a 2.5% increase to potatoes, OR a winter frog changes all the trees on your land to leafless trees, OR, a rain forest frog brings constant rain to your land, these are just a few ideas. Honestly, not everything needs utility, SFL has so many utility items, sometimes it's just fun to collect unique things, especially when we can show it off, decor, avatar photo, etc.

**Comment by Yougottabekiddingme123** [comment]
Not everyone thinks like this. Just look at the t2 beaver, the demand for
them is extremely low and had to be artificially  increased with quest
utility.

Utility is very important in this game. Without it the demand for frogs
will also be low.

**Comment by lelduck** [comment]
Idea for utility. expand land crops onto water crops in the pond, the frog enables boosts to water crops or the frog itself unlock water crops

**Reply by AL-Simmons92** [reply]
They have sooo many options. more crops....1 million options for crops, fruit tress, nut trees, water plans.... herb.....etc etc.... wild life...

**Comment by alphacruze** [comment]
Why such a low limit of Frogs? I mean this a game, not a PFP collection where a super low supply is best
